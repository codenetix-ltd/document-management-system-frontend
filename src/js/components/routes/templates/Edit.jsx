import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import autobind from 'autobind-decorator';

import ContentHeader from 'Components/ContentHeader';
import ContentWrapper from 'Components/ContentWrapper';
import TemplateForm from 'Routes/templates/partials/Form';

import AlertMessage from 'Components/common/AlertMessage';
import ErrorMessage from 'Components/common/ErrorMessage';

@autobind
export class TemplateEdit extends Component {
  static propTypes = {
    match: PropTypes.any.isRequired,
    template: PropTypes.any.isRequired
  };

  render() {
    const breadcrumbs = [
      { pageName: 'Templates', pageLink: '/templates/list', iconCls: 'fa fa-copy' },
      { pageName: 'Edit template', pageLink: '/templates', iconCls: 'fa fa-pencil' }
    ];

    return (
      <div>
        <ContentHeader title={`Edit ${this.props.template.name}`} breadcrumbs={breadcrumbs} />
        <ContentWrapper boxClass="box-info">
          <div className="box-header with-border">
            <h3 className="box-title">Template data</h3>
          </div>
          <div className="box-body">
            <div className="row">
              <div className="col-sm-12">
                <AlertMessage />
                <ErrorMessage />
              </div>
            </div>
          </div>
          <TemplateForm
            match={this.props.match}
            submitButtonText="Update"
          />
        </ContentWrapper>
      </div>
    );
  }
}

const mapStateToProps = ({ template }) => ({ template });

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(TemplateEdit);
