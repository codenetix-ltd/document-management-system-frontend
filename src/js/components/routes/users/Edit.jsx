import React, { Component } from 'react';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';

import ContentHeader from 'Components/ContentHeader';
import ContentWrapper from 'Components/ContentWrapper';
import UserForm from 'Routes/users/partials/Form';

import axios from 'Services/request';
import { API } from 'Config';

@autobind
export default class UserEdit extends Component {
  static propTypes = {
    match: PropTypes.any.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      user: {}
    };
  }

  componentDidMount() {
    this.fetchUser();
  }

  onFormSubmit(formData) {
    const { userID } = this.props.match.params;
    axios.put(`${API.users}/${userID}`, formData).catch(console.trace);
  }

  getTemplateOptions(input, callback) {
    this.fetchTemplates().then(({ data }) => {
      const list = data.data.map(({ id, name }) => ({ label: name, value: id }));
      callback(null, {
        options: list,
        complete: true
      });
    }).catch(callback);
  }

  getRoleOptions(input, callback) {
    this.fetchRoles().then(({ data }) => {
      const list = data.data.map(({ id, name }) => ({ label: name, value: id }));
      callback(null, {
        options: list,
        complete: true
      });
    }).catch(callback);
  }

  fetchTemplates() {
    return axios.get(API.templates);
  }

  fetchRoles() {
    return axios.get(API.roles);
  }

  // todo: move this to a thunk
  fetchUser() {
    const { userID } = this.props.match.params;
    axios.get(`${API.users}/${userID}`).then(({ data }) => {
      const {
        fullName,
        email,
        templatesIds,
        ...rest
      } = data;
      this.fetchTemplates().then(res => {
        const options = res.data.data.map(({ id, name }) => ({ label: name, value: id }));
        const selected = options.filter(option => templatesIds.includes(option.value));
        this.setState({
          user: {
            fullName,
            email,
            ...rest,
            templatesIds: selected
          }
        });
      });
    }).catch(err => {
      throw err;
    });
  }

  breadcrumbs = [
    { pageName: 'Users', pageLink: '/users', iconCls: 'fa fa-users' }
  ];

  render() {
    return (
      <div>
        <ContentHeader title={`Edit ${this.state.user.fullName || ''}`} breadcrumbs={this.breadcrumbs} />
        <ContentWrapper boxClass="box-info">
          <div className="box-header with-border">
            <h3 className="box-title">Profile data</h3>
          </div>
          <UserForm
            user={this.state.user}
            onSubmit={this.onFormSubmit}
            getTemplateOptions={this.getTemplateOptions}
            getRoleOptions={this.getRoleOptions}
            submitButtonText="Update"
          />
        </ContentWrapper>
      </div>
    );
  }
}
