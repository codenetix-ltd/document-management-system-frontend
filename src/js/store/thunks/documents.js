import { API } from 'Config';
import pick from 'lodash/pick';
import pickBy from 'lodash/pickBy';
import identity from 'lodash/identity';

import axios from 'Services/request';

import {
  $loading,
  $document,
  $updateDocument,
  $documentsList,
  $setComparedDocuments
} from 'Store/actions';

import { $$messageSet } from 'Store/thunks/message';
import { $$errorSet } from 'Store/thunks/error';

import initialState from 'Store/reducers/initialState.json';

const fn = () => {};

export const $$documentsFetch = (dispatch, params, callback = fn) => {
  const picked = pick(params, ['page', 'sortField', 'sortDirection']);
  const filters = pickBy(params.filterSet, identity);
  dispatch($loading(true));
  axios.get(API.documents, {
    params: { ...picked, ...filters }
  }).then(({ data }) => {
    dispatch($documentsList({
      list: data.data,
      lastPage: data.meta.lastPage,
      ...picked
    }));
    dispatch($loading(false));
    callback(data);
  }).catch(err => {
    console.trace(err);
    dispatch($loading(false));
    $$errorSet(dispatch, err);
  });
};

export const $$documentFetch = (dispatch, documentID, callback = fn) => {
  dispatch($loading(true));
  axios.get(`${API.documents}/${documentID}`).then(({ data }) => {
    dispatch($document({ ...data }));
    dispatch($loading(false));
    callback(data);
  }).catch(err => {
    console.trace(err);
    dispatch($loading(false));
    $$errorSet(dispatch, err);
  });
};

export const $$documentUpdate = (dispatch, part) => {
  dispatch($updateDocument(part));
};

export const $$documentReset = (dispatch) => {
  dispatch($document(initialState.document));
};

export const $$documentDelete = (dispatch, { id, name }, close = fn) => {
  axios.delete(`${API.documents}/${id}`).then(() => {
    $$documentsFetch(dispatch, 1);
    $$messageSet(dispatch, {
      type: 'success',
      text: `The document ${name} was successfully deleted.`
    });
    close();
  }).catch(err => {
    console.trace(err);
    close();
    $$errorSet(dispatch, err);
  });
};

export const $$documentsMassDelete = (dispatch, ids, close = fn) => {
  const p = Promise.all(ids.map(id => axios.delete(`${API.documents}/${id}`)));
  p.then(() => {
    $$documentsFetch(dispatch, 1);
    $$messageSet(dispatch, {
      type: 'success',
      text: 'Selected documents were successfully deleted.'
    });
    close();
  }).catch(err => {
    console.trace(err);
    close();
    $$errorSet(dispatch, err);
  });
};

export const $$comparedDocumentsFetch = (dispatch, ids, callback = fn) => {
  const p = Promise.all(ids.map(id => axios.get(`${API.documents}/${id}`)));
  p.then(d => {
    dispatch($setComparedDocuments(d.map(({ data }) => data)));
    callback(d);
  }).catch(err => {
    console.trace(err);
  });
};
