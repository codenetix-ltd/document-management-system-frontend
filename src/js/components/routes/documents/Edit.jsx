import axios from 'axios';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { If, Then } from 'qc-react-conditionals/lib';
import { Tabs, Tab } from 'react-bootstrap';

import ContentHeader from 'Components/ContentHeader';
import ContentWrapper from 'Components/ContentWrapper';

import { API } from 'Config';

import { $$documentFetch } from 'Store/thunks/documents';

import DocumentForm from './partials/Form';
import AttributesForm from './partials/AttributesForm';
import DocumentVersions from './partials/DocumentVersions';

@autobind
export class DocumentEdit extends Component {
  static propTypes = {
    match: PropTypes.any.isRequired,
    dispatch: PropTypes.func.isRequired,
    document: PropTypes.any.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      activeKey: 1,
      newDocumentID: null
    };
    this.breadcrumbs = [
      { pageName: 'Documents', pageLink: '/documents/list', iconCls: 'fa fa-copy' },
      { pageName: 'Edit document', pageLink: '', iconCls: 'fa fa-pencil' }
    ];
  }

  componentDidMount() {
    const { dispatch, match } = this.props;
    const { documentID } = match.params;
    $$documentFetch(dispatch, documentID);
  }

  onTabSelect(activeKey) {
    this.setState({ activeKey });
  }

  onFormSubmit() {
    const { document, match } = this.props;
    const { documentID } = match.params;
    if (documentID) {
      axios.put(API.documents, document).catch(e => {
        console.trace(e);
      });
    } else {
      axios.post(API.documents, document).then(({ data }) => {
        if (!data.id) throw new Error('No id field in response.');
        this.setState({ newDocumentID: data.id });
      }).catch(e => {
        console.trace(e);
      });
    }
  }

  validate() {
    const { document } = this.props;
    return !Object.keys(document).every(key => !!document[key]);
  }

  render() {
    const { activeKey, newDocumentID } = this.state;
    const { document: { actualVersion }, match, match: { params: documentID } } = this.props;
    if (newDocumentID) {
      return <Redirect to={`/documents/${newDocumentID}`} />;
    }
    return (
      <div>
        <ContentHeader title={`Edit document ${actualVersion.name}`} breadcrumbs={this.breadcrumbs} />
        <ContentWrapper noBox>
          <Tabs id="tabs" activeKey={activeKey} onSelect={this.onTabSelect} animation={false}>
            <Tab eventKey={1} title="Content">
              <div className="box-header with-border">
                <h3 className="box-title">General information</h3>
              </div>
              <DocumentForm />
              <If is={actualVersion.template}>
                <Then>
                  <div className="box-header with-border">
                    <h3 className="box-title">Document attributes</h3>
                  </div>
                  <AttributesForm />
                </Then>
              </If>
              <div className="box-footer">
                <button
                  type="button"
                  className="btn btn-success"
                  onClick={this.onFormSubmit}
                  disabled={this.validate()}
                >
                  { documentID ? 'Update' : 'Create' }
                </button>
              </div>
            </Tab>
            <Tab eventKey={2} title="Versions">
              <DocumentVersions match={match} />
            </Tab>
          </Tabs>
        </ContentWrapper>
      </div>
    );
  }
}

const mapStateToProps = ({ document }) => ({ document });

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(DocumentEdit);
