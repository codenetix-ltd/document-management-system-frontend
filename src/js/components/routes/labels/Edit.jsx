import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';

import ContentHeader from 'Components/ContentHeader';
import ContentWrapper from 'Components/ContentWrapper';
import LabelForm from 'Routes/labels/partials/Form';

import axios from 'Services/request';
import { API } from 'Config';

import { $$labelFetch } from 'Store/thunks/labels';

@autobind
export class LabelEdit extends Component {
  static propTypes = {
    match: PropTypes.any.isRequired,
    dispatch: PropTypes.func.isRequired,
    label: PropTypes.any.isRequired
  };

  componentDidMount() {
    this.fetchLabel();
  }

  onFormSubmit(formData) {
    const { labelID } = this.props.match.params;
    axios.put(`${API.labels}/${labelID}`, formData).catch(err => {
      throw err;
    });
  }

  fetchLabel() {
    const { dispatch } = this.props;
    const { labelID } = this.props.match.params;
    $$labelFetch(dispatch, labelID);
  }

  breadcrumbs = [
    { pageName: 'Labels', pageLink: '/labels', iconCls: 'fa fa-tags' }
  ];

  render() {
    const { label } = this.props;
    return (
      <div>
        <ContentHeader title={`Edit ${label.name || ''}`} breadcrumbs={this.breadcrumbs} />
        <ContentWrapper boxClass="box-info">
          <div className="box-header with-border">
            <h3 className="box-title">Label</h3>
          </div>
          <LabelForm
            label={label}
            onSubmit={this.onFormSubmit}
            submitButtonText="Update"
          />
        </ContentWrapper>
      </div>
    );
  }
}

const mapStateToProps = ({ label }) => ({ label });

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(LabelEdit);
