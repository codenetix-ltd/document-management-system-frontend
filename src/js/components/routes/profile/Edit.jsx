import axios from 'axios';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';

import ContentHeader from 'Components/ContentHeader';
import ContentWrapper from 'Components/ContentWrapper';
import ProfileForm from 'Routes/profile/partials/Form';

import { API } from 'Config';

import { $$profileFetch } from 'Store/thunks/profile';

@autobind
export class ProfileEdit extends Component {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    profile: PropTypes.any.isRequired
  };

  constructor(props) {
    super(props);
    this.breadcrumbs = [
      { pageName: 'My profile', pageLink: '/profile', iconCls: '' }
    ];
  }

  componentDidMount() {
    $$profileFetch(this.props.dispatch);
  }

  onFormSubmit() {
    const formData = this.props.profile;
    axios.put(`${API.profile}`, formData).catch(err => {
      throw err;
    });
  }

  render() {
    return (
      <div>
        <ContentHeader title="My profile" breadcrumbs={this.breadcrumbs} />
        <ContentWrapper boxClass="box-info">
          <div className="box-header with-border">
            <h3 className="box-title">Profile data</h3>
          </div>
          <ProfileForm onSubmit={this.onFormSubmit} />
        </ContentWrapper>
      </div>
    );
  }
}

const mapStateToProps = ({ profile }) => ({ profile });

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(ProfileEdit);
