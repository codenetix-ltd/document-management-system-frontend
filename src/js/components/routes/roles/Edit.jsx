import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';

import ContentHeader from 'Components/ContentHeader';
import ContentWrapper from 'Components/ContentWrapper';
import RoleForm from 'Routes/roles/partials/Form';

import axios from 'Services/request';
import { API } from 'Config';

import { $$roleFetch } from 'Store/thunks/roles';

@autobind
export class RoleEdit extends Component {
  static propTypes = {
    match: PropTypes.any.isRequired,
    dispatch: PropTypes.func.isRequired,
    role: PropTypes.any.isRequired
  };

  componentDidMount() {
    this.fetchRole();
  }

  onFormSubmit(formData) {
    const { roleID } = this.props.match.params;
    axios.put(`${API.roles}/${roleID}`, formData).catch(err => {
      throw err;
    });
  }

  fetchRole() {
    const { dispatch } = this.props;
    const { roleID } = this.props.match.params;
    $$roleFetch(dispatch, roleID);
  }

  render() {
    const breadcrumbs = [
      { pageName: 'Roles', pageLink: '/roles', iconCls: 'fa fa-tags' }
    ];
    const { role } = this.props;
    return (
      <div>
        <ContentHeader title={`Edit ${role.name || ''}`} breadcrumbs={breadcrumbs} />
        <ContentWrapper boxClass="box-info">
          <div className="box-header with-border">
            <h3 className="box-title">Role</h3>
          </div>
          <div className="box-body">
            <RoleForm
              role={role}
              onSubmit={this.onFormSubmit}
              submitButtonText="Update"
            />
          </div>
        </ContentWrapper>
      </div>
    );
  }
}

const mapStateToProps = ({ role }) => ({ role });

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(RoleEdit);
