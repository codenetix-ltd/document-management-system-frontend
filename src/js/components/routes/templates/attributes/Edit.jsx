import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import autobind from 'autobind-decorator';

import ContentHeader from 'Components/ContentHeader';
import ContentWrapper from 'Components/ContentWrapper';
import AttributeForm from 'Routes/templates/attributes/partials/Form';

import AlertMessage from 'Components/common/AlertMessage';
import ErrorMessage from 'Components/common/ErrorMessage';

import { $$attributeFetch } from 'Store/thunks/attributes';

import axios from 'Services/request';
import { getURL } from 'Config';

@autobind
export class AttributeEdit extends Component {
  static propTypes = {
    match: PropTypes.any.isRequired,
    attribute: PropTypes.any.isRequired,
    dispatch: PropTypes.any.isRequired
  };

  componentDidMount() {
    this.fetchAttribute();
  }

  onFormSubmit(formData) {
    const { templateID, attributeID } = this.props.match.params;
    const url = getURL('attributes', templateID);
    axios.put(`${url}/${attributeID}`, formData).catch(console.trace);
  }

  getBreadcrumbs(templateID) {
    return [
      { pageName: 'Templates', pageLink: '/templates/list', iconCls: 'fa fa-copy' },
      { pageName: `Edit template ${templateID}`, pageLink: `/templates/${templateID}`, iconCls: 'fa fa-pencil' },
      { pageName: 'Template attribute edit', pageLink: '', iconCls: 'fa fa-plus' }
    ];
  }

  fetchAttribute() {
    const { dispatch, match } = this.props;
    const { templateID, attributeID } = match.params;
    $$attributeFetch(dispatch, templateID, attributeID);
  }

  render() {
    const { templateID } = this.props.match.params;
    return (
      <div>
        <ContentHeader title={`Edit ${this.props.attribute.name}`} breadcrumbs={this.getBreadcrumbs(templateID)} />
        <ContentWrapper boxClass="box-info">
          <div className="box-header with-border">
            <h3 className="box-title">Attribute data</h3>
          </div>
          <div className="box-body">
            <div className="row">
              <div className="col-sm-12">
                <AlertMessage />
                <ErrorMessage />
              </div>
            </div>
          </div>
          <AttributeForm
            onSubmit={this.onFormSubmit}
            submitButtonText="Update"
          />
        </ContentWrapper>
      </div>
    );
  }
}

const mapStateToProps = ({ attribute }) => ({ attribute });

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(AttributeEdit);
