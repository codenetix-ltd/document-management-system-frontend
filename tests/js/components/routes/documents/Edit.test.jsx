import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { DocumentEdit } from 'Routes/documents/Edit';

import { $$documentFetch } from 'Store/thunks/documents';

jest.mock('Store/thunks/documents', () => ({
  $$documentFetch: jest.fn()
}));

describe('Documents Edit', () => {
  let wrapper;
  let match;
  let props;
  let instance;

  beforeEach(() => {
    match = {
      params: {
        documentID: 1
      }
    };
    props = {
      match,
      dispatch: () => {},
      document: {
        actualVersion: {}
      }
    };
    wrapper = shallow(<DocumentEdit {...props} />);
    instance = wrapper.instance();
  });

  it('should render correctly', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should call $$documentFetch in componentDidMount', () => {
    expect($$documentFetch.mock.calls.length).toBe(1);
  });

  it('should change state upon calling onTabSelect', () => {
    expect(instance.state).toHaveProperty('activeKey');
    expect(instance.state.activeKey).toBe(1);
    instance.onTabSelect(2);
    expect(instance.state.activeKey).toBe(2);
  });

  it('should call validate method', () => {
    const doc = {
      name: ''
    };
    expect(instance.validate(doc)).toBe(false);
    doc.name = 'Lorem ipsum';
    expect(instance.validate(doc)).toBe(true);
  });
});
