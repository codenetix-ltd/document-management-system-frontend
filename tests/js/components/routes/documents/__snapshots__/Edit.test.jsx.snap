// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Documents Edit should render correctly 1`] = `
<div>
  <ContentHeader
    breadcrumbs={
      Array [
        Object {
          "iconCls": "fa fa-copy",
          "pageLink": "/documents/list",
          "pageName": "Documents",
        },
        Object {
          "iconCls": "fa fa-pencil",
          "pageLink": "",
          "pageName": "Edit document",
        },
      ]
    }
    title="Edit document undefined"
  />
  <ContentWrapper
    boxClass=""
    noBox={true}
  >
    <Uncontrolled(Tabs)
      activeKey={1}
      animation={false}
      id="tabs"
      onSelect={[Function]}
    >
      <Tab
        eventKey={1}
        title="Content"
      >
        <div
          className="box-header with-border"
        >
          <h3
            className="box-title"
          >
            General information
          </h3>
        </div>
        <Connect(DocumentForm) />
        <If>
          <Then>
            <div
              className="box-header with-border"
            >
              <h3
                className="box-title"
              >
                Document attributes
              </h3>
            </div>
            <Connect(AttributesForm) />
          </Then>
        </If>
        <div
          className="box-footer"
        >
          <button
            className="btn btn-success"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Update
          </button>
        </div>
      </Tab>
      <Tab
        eventKey={2}
        title="Versions"
      >
        <Connect(DocumentVersions)
          match={
            Object {
              "params": Object {
                "documentID": 1,
              },
            }
          }
        />
      </Tab>
    </Uncontrolled(Tabs)>
  </ContentWrapper>
</div>
`;
